Add policy to find adjacent nodes by selecting node

crystalmdsouza opened the pull request:
The NodesAndAdjacentNodes policy highlights the adjacent nodes (those which are connected by a linked edge) of an inspected/selected node. A single selection is set with all adjacent nodes and the inspected/selected node, with the inspected/selected node as the last entry to identify it from its adjacent nodes. If a node does not have any adjacent nodes, only the inspected/selected node is set in the selection.

I added a brief explanation of the policy that gets pulled into the reference guide, and an associated unit test similar to those for the NodesAndLinkedEdges and EdgesAndLinkedNodes selection policies.

- [x] issues: fixes #11806 
- [x] tests added / passed
- [x] release document entry (if new feature or API change)

bryevdv commented:
Note that the codebase tests failures are repo-wide and unrelated to your code. We will have a fix pushed soon and then PRs will need to rebase/merge main in order to go green

crystalmdsouza commented:
@bryevdv Sounds good, thanks for the update!

bryevdv commented on bokeh/models/graphs.py:
```diff
@@ -161,6 +162,19 @@ class EdgesAndLinkedNodes(GraphHitTestPolicy):
+    those directly connected to it by edges. There is no direct selection
+    or inspection of graph edges or nodes. The last entry in each list
+    returned by NodesAndAdjacentNodes refers to the original node that was
+    inspected or selected.
```
Can we contrive it to always be the first, instead? I think that would be simpler for users to make use of.

crystalmdsouza replied:
Initially I had set the selected node to be the first in the list. However, I noticed in bokeh/examples/plotting/file/graphs.py that when the hover tool is set to print selection data in the graphic, it prints the data over the last node in the selection list. See the attached video for reference. (https://user-images.githubusercontent.com/90462878/145235745-1f37ce8a-7f92-40e2-bcec-ec42fb719a4a.mov).

I'm not too familiar with the hover tool or the mechanisms behind it, but if there is a way to make it use the first element instead of the last as an indicator of where to print selection data, then I'd be happy to make this change!

bryevdv replied:
Oh I see, I am not sure we have ever actually made any specific guarantees about the order. If it is not something that is documented and under test, this its just be undefined behavior that happens to be useful. I think it would be safest to just remove the claim here altogether for now, and make a separate issues about finding a way to communicate the "tool selected" nodes vs the "policy selected" node. Then a PR can consider all the Polices together (to take a consistent approach) and also make sure everything is documented together and maintained under test.

crystalmdsouza replied:
Sure, I'll edit the claim accordingly in the policy description!

bryevdv commented on bokeh/models/graphs.py:
```diff
@@ -161,6 +162,19 @@ class EdgesAndLinkedNodes(GraphHitTestPolicy):
 
+class NodesAndAdjacentNodes(GraphHitTestPolicy):
+    '''
+    With the ``NodesAndAdjacentNodes`` policy, inspection or selection of
+    graph nodes will result in the inspection or selection of the node and
+    those directly connected to it by edges. There is no direct selection
+    or inspection of graph edges or nodes. The last entry in each list
```
This wording is a little confusing to me. Maybe something like:
```suggestion
    With the ``NodesAndAdjacentNodes`` policy, inspection or selection of
    graph nodes will also result in the inspection or selection any nodes that
    are immediately adjacent (connected by a single edge). There is no selection
    or inspection of graph edges. The last entry in each list
```

cc @tcmetzger for suggestions also

crystalmdsouza replied:
Your version does sound more readable, I will change the description accordingly. :)

bryevdv commented on bokehjs/src/lib/models/graphs/graph_hit_test_policy.ts:
```diff
@@ -283,3 +283,85 @@ export class EdgesAndLinkedNodes extends GraphHitTestPolicy {
+    const initial_node_selection = graph.node_renderer.data_source.selected
+    initial_node_selection.update(hit_test_result, final, mode)
+
+    const final_node_selection = graph.node_renderer.data_source.selected
```
I had to squint at this for a second. I think assigning `graph.node_renderer.data_source.selected` to two different names, which are then used to mutate the same value, is confusing. I'd prefer to re-work this with a single
```
const node_selection = graph.node_renderer.data_source.selected
``` 
at the top that is used throughout

bryevdv replied:
Or if you want to be more compact
```
const {selected} = graph.node_renderer.data_source
```
that is used throughout for updates

crystalmdsouza replied:
Sure, I can change this to be more readable!

g-parki commented:
This is cool! Thanks for the contribution @crystalmdsouza 

https://user-images.githubusercontent.com/61096711/145152480-000a4a9f-050c-4314-a891-5f8f2cfc946a.mov

bryevdv commented:
@crystalmdsouza this is looking good, I only have the one suggestion left about the result order claim. Other last things:

* can you also add a bullet point at the top of https://github.com/bokeh/bokeh/blob/branch-3.0/sphinx/source/docs/releases/3.0.0.rst ? We always do a final edit of the release notes, so just a quick placeholder sentence will do

* In case you want a "stretch" goal: it would be nice to update example that @g-parki modified above to have a dropdown that allows selecting and trying out all the different selection policies. Definitely an optional ask, up to you in case you have the bandwidth to look at it.

g-parki commented:
Just FYI the example in the demo is [graphs.py](https://github.com/bokeh/bokeh/blob/branch-3.0/examples/plotting/file/graphs.py)

bryevdv commented:
I guess I should add: if you do want to take a look at the example, there's at least some chance that swapping out policies dynamically doesn't actually work (AFAIK no one's ever tried) so if a quick attempt doesn't function as expected, we can just punt on that and make a bug report.

crystalmdsouza commented:
@bryevdv Thanks for the follow up! I've added the bullet point to the release notes and updated the claim about results ordering, but dynamically changing the selection policy via a dropdown didn't work though I tried it a few different ways. 

Could I instead add two graphs at the end of this example that use the NodesAndAdjacentNodes inspection/selection policy to give users the chance to see all three (NodesAndLinkedEdges, EdgesAndLinkedNodes, NodesAndAdjacentNodes) in action?

bryevdv commented:
> Could I instead add two graphs at the end of this example 

Certainly that's fine. My suggesting with the dropdown was in order to try keep things a bit more compact. Alternatively I wonder if it would be  possible to put the plots in to tabbed sphinx-panels instead cc @tcmetzger for thoughts

tcmetzger commented:
@crystalmdsouza Thanks for your contribution, that looks great! We are using https://sphinx-tabs.readthedocs.io/en/latest/ in our docs, so this might work with RST, at least.

crystalmdsouza commented:
Thanks @bryevdv and @tcmetzger for the guidance! I have added examples of the NodesAndLinkedEdges, EdgesAndLinkedNodes, and NodesAndAdjacentNodes policies into the graphs documentation rst, and added two graphs showing the NodesAndAdjacentNodes policy in bokeh/examples/models/file/graphs.py. Please let me know if there is anything else I should change!

Also, I noticed that the BokehJS-CI/test for "windows-latest" has been failing for this pull request recently, is this something that I should be concerned with? I don't really understand what it means.

g-parki commented:
> BokehJS-CI/test for "windows-latest" has been failing for this pull request recently, is this something that I should be concerned with?

This looks like another repo-wide problem, so I submitted an issue for it. You may have to rebase again once it's cleared up.

bryevdv commented:
Thanks @crystalmdsouza !

crystalmdsouza commented:
No problem! Thank you @bryevdv, @g-parki and @tcmetzger for your help!
